### ISSUE TITLE
Evaluation doesn't produce same result as direct applying of loss function when padding is used

### ISSUE THREAD

**Issue by danpetruk** [keras-team/keras#14899]
**System information**.
- Have I written custom code (as opposed to using a stock example script provided in Keras):
- Windows 10
- TensorFlow v2.5.0-rc3-213-ga4dfb8d1a71 2.5.0 (from pip)
- Python 3.8.0 (tags/v3.8.0:fa919fd, Oct 14 2019, 19:37:50) [MSC v.1916 64 bit (AMD64)]
- GTX 1050 2gb

**Describe the problem**.

`model.evaluate(x,y)` and `loss_function(y, model.predict(x))` return different results although they're supposed to return same result

**Source code / logs**.

```
from tensorflow import keras
from tensorflow.keras import layers
import numpy as np

input = layers.Input((None,1))
mask = layers.Masking(mask_value=1,dtype="float32")(input)
output = layers.TimeDistributed(layers.Dense(6,activation="softmax"))(mask)
model = keras.Model(inputs=input, outputs=output)
model.compile(loss="mse")

x=tf.keras.preprocessing.sequence.pad_sequences([np.array([[0.]]),np.array([[0.],[0.]])],value=1,padding="post",dtype="float32")
y=tf.keras.preprocessing.sequence.pad_sequences([np.array([[0.5,0.5,0.5,0.5,0.5,0.5]]),np.array([[0.5,0.5,0.5,0.5,0.5,0.5],[0.5,0.5,0.5,0.5,0.5,0.5]])],value=1,padding="post",dtype="float32")

print(model.evaluate(x,y)) #0.083333320915699

mse = tf.keras.losses.MeanSquaredError()
print(mse(y, model.predict(x)).numpy()) #0.25694442

```

**Comment by rmothukuru** [comment]
Could reproduce the issue in **`Google Colab`** with **`Tensorflow Version 2.5`**. Please find [the Gist](https://colab.research.google.com/gist/rmothukuru/fec1e855dd66ae5934105dc7095dd99a/keras_14899.ipynb).

**Comment by LukeWood** [comment]
Hey there @danpetruk, would you be interested in contributing a fix for this?

**Comment by danpetruk** [comment]
Hey @LukeWood, thanks for a suggestion, but unfortunately I don't have neither enough skill not time for this

**Comment by LukeWood** [comment]
thanks, re-added the keras-team-review-pending label.

**Comment by yashk2810** [comment]
Triage: This is happening because of the Mask layer which is taken into account when calling model.evaluate as opposed to just calling the loss function.

**Comment by qlzh727** [comment]
I think this issue here is because of masking value for the inputs.

When model.eval is called, any metric/loss calculation will ignore any masked value in the input, and only report the value for original data. However, when you do `mse(y, model.predict(x))`, the metric instance doesn't aware any masking information, and take all the data into account, which is why it report different value.

If you want the mse to aware the masking, you could use `sample_weight` params when calling the metric. See  https://keras.io/api/metrics/accuracy_metrics/ for how to mask values in metric class.
